File: go/pkg/basecamp/forwards.go
package basecamp

import (
	"context"
	"fmt"
	"time"

	"github.com/basecamp/basecamp-sdk/go/pkg/generated"
)

// Inbox represents a Basecamp email inbox (forwards tool).
type Inbox struct {
	ID        int64     `json:"id"`
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Title     string    `json:"title"`
	Type      string    `json:"type"`
	URL       string    `json:"url"`
	AppURL    string    `json:"app_url"`
	Bucket    *Bucket   `json:"bucket,omitempty"`
	Creator   *Person   `json:"creator,omitempty"`
}

// Forward represents a forwarded email in Basecamp.
type Forward struct {
	ID        int64     `json:"id"`
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Subject   string    `json:"subject"`
	Content   string    `json:"content"`
	From      string    `json:"from"`
	Type      string    `json:"type"`
	URL       string    `json:"url"`
	AppURL    string    `json:"app_url"`
	Parent    *Parent   `json:"parent,omitempty"`
	Bucket    *Bucket   `json:"bucket,omitempty"`
	Creator   *Person   `json:"creator,omitempty"`
}

// ForwardReply represents a reply to a forwarded email.
type ForwardReply struct {
	ID        int64     `json:"id"`
	Status    string    `json:"status"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Content   string    `json:"content"`
	Type      string    `json:"type"`
	URL       string    `json:"url"`
	AppURL    string    `json:"app_url"`
	Parent    *Parent   `json:"parent,omitempty"`
	Bucket    *Bucket   `json:"bucket,omitempty"`
	Creator   *Person   `json:"creator,omitempty"`
}

// CreateForwardReplyRequest specifies the parameters for creating a reply to a forward.
type CreateForwardReplyRequest struct {
	// Content is the reply body in HTML (required).
	Content string `json:"content"`
}

// ForwardsService handles email forward operations.
type ForwardsService struct {
	client *AccountClient
}

// NewForwardsService creates a new ForwardsService.
func NewForwardsService(client *AccountClient) *ForwardsService {
	return &ForwardsService{client: client}
}

// GetInbox returns an inbox by ID.
// bucketID is the project ID, inboxID is the inbox ID.
func (s *ForwardsService) GetInbox(ctx context.Context, bucketID, inboxID int64) (result *Inbox, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "GetInbox",
		ResourceType: "inbox", IsMutation: false,
		BucketID: bucketID, ResourceID: inboxID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.GetInboxWithResponse(ctx, s.client.accountID, bucketID, inboxID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		err = fmt.Errorf("unexpected empty response")
		return nil, err
	}

	inbox := inboxFromGenerated(resp.JSON200.Inbox)
	return &inbox, nil
}

// List returns all forwards in an inbox.
// bucketID is the project ID, inboxID is the inbox ID.
func (s *ForwardsService) List(ctx context.Context, bucketID, inboxID int64) (result []Forward, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "List",
		ResourceType: "forward", IsMutation: false,
		BucketID: bucketID, ResourceID: inboxID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.ListForwardsWithResponse(ctx, s.client.accountID, bucketID, inboxID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, nil
	}

	forwards := make([]Forward, 0, len(*resp.JSON200))
	for _, gf := range *resp.JSON200 {
		forwards = append(forwards, forwardFromGenerated(gf))
	}

	return forwards, nil
}

// Get returns a forward by ID.
// bucketID is the project ID, forwardID is the forward ID.
func (s *ForwardsService) Get(ctx context.Context, bucketID, forwardID int64) (result *Forward, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "Get",
		ResourceType: "forward", IsMutation: false,
		BucketID: bucketID, ResourceID: forwardID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.GetForwardWithResponse(ctx, s.client.accountID, bucketID, forwardID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		err = fmt.Errorf("unexpected empty response")
		return nil, err
	}

	forward := forwardFromGenerated(resp.JSON200.Forward)
	return &forward, nil
}

// ListReplies returns all replies to a forward.
// bucketID is the project ID, forwardID is the forward ID.
func (s *ForwardsService) ListReplies(ctx context.Context, bucketID, forwardID int64) (result []ForwardReply, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "ListReplies",
		ResourceType: "forward_reply", IsMutation: false,
		BucketID: bucketID, ResourceID: forwardID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.ListForwardRepliesWithResponse(ctx, s.client.accountID, bucketID, forwardID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		return nil, nil
	}

	replies := make([]ForwardReply, 0, len(*resp.JSON200))
	for _, gr := range *resp.JSON200 {
		replies = append(replies, forwardReplyFromGenerated(gr))
	}

	return replies, nil
}

// GetReply returns a forward reply by ID.
// bucketID is the project ID, forwardID is the forward ID, replyID is the reply ID.
func (s *ForwardsService) GetReply(ctx context.Context, bucketID, forwardID, replyID int64) (result *ForwardReply, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "GetReply",
		ResourceType: "forward_reply", IsMutation: false,
		BucketID: bucketID, ResourceID: replyID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	resp, err := s.client.gen.GetForwardReplyWithResponse(ctx, s.client.accountID, bucketID, forwardID, replyID)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		err = fmt.Errorf("unexpected empty response")
		return nil, err
	}

	reply := forwardReplyFromGenerated(resp.JSON200.Reply)
	return &reply, nil
}

// CreateReply creates a new reply to a forwarded email.
// bucketID is the project ID, forwardID is the forward ID.
// Returns the created reply.
func (s *ForwardsService) CreateReply(ctx context.Context, bucketID, forwardID int64, req *CreateForwardReplyRequest) (result *ForwardReply, err error) {
	op := OperationInfo{
		Service: "Forwards", Operation: "CreateReply",
		ResourceType: "forward_reply", IsMutation: true,
		BucketID: bucketID, ResourceID: forwardID,
	}
	if gater, ok := s.client.parent.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.parent.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.parent.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	if req == nil || req.Content == "" {
		err = ErrUsage("reply content is required")
		return nil, err
	}

	body := generated.CreateForwardReplyJSONRequestBody{
		Content: req.Content,
	}

	resp, err := s.client.gen.CreateForwardReplyWithResponse(ctx, s.client.accountID, bucketID, forwardID, body)
	if err != nil {
		return nil, err
	}
	if err = checkResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	if resp.JSON200 == nil {
		err = fmt.Errorf("unexpected empty response")
		return nil, err
	}

	reply := forwardReplyFromGenerated(resp.JSON200.Reply)
	return &reply, nil
}

// inboxFromGenerated converts a generated Inbox to our clean type.
func inboxFromGenerated(gi generated.Inbox) Inbox {
	i := Inbox{
		Status:    gi.Status,
		CreatedAt: gi.CreatedAt,
		UpdatedAt: gi.UpdatedAt,
		Title:     gi.Title,
		Type:      gi.Type,
		URL:       gi.Url,
		AppURL:    gi.AppUrl,
	}

	if gi.Id != nil {
		i.ID = *gi.Id
	}

	if gi.Bucket.Id != nil || gi.Bucket.Name != "" {
		i.Bucket = &Bucket{
			ID:   derefInt64(gi.Bucket.Id),
			Name: gi.Bucket.Name,
			Type: gi.Bucket.Type,
		}
	}

	if gi.Creator.Id != nil || gi.Creator.Name != "" {
		i.Creator = &Person{
			ID:           derefInt64(gi.Creator.Id),
			Name:         gi.Creator.Name,
			EmailAddress: gi.Creator.EmailAddress,
			AvatarURL:    gi.Creator.AvatarUrl,
			Admin:        gi.Creator.Admin,
			Owner:        gi.Creator.Owner,
		}
	}

	return i
}

// forwardFromGenerated converts a generated Forward to our clean type.
func forwardFromGenerated(gf generated.Forward) Forward {
	f := Forward{
		Status:    gf.Status,
		CreatedAt: gf.CreatedAt,
		UpdatedAt: gf.UpdatedAt,
		Subject:   gf.Subject,
		Content:   gf.Content,
		From:      gf.From,
		Type:      gf.Type,
		URL:       gf.Url,
		AppURL:    gf.AppUrl,
	}

	if gf.Id != nil {
		f.ID = *gf.Id
	}

	if gf.Parent.Id != nil || gf.Parent.Title != "" {
		f.Parent = &Parent{
			ID:     derefInt64(gf.Parent.Id),
			Title:  gf.Parent.Title,
			Type:   gf.Parent.Type,
			URL:    gf.Parent.Url,
			AppURL: gf.Parent.AppUrl,
		}
	}

	if gf.Bucket.Id != nil || gf.Bucket.Name != "" {
		f.Bucket = &Bucket{
			ID:   derefInt64(gf.Bucket.Id),
			Name: gf.Bucket.Name,
			Type: gf.Bucket.Type,
		}
	}

	if gf.Creator.Id != nil || gf.Creator.Name != "" {
		f.Creator = &Person{
			ID:           derefInt64(gf.Creator.Id),
			Name:         gf.Creator.Name,
			EmailAddress: gf.Creator.EmailAddress,
			AvatarURL:    gf.Creator.AvatarUrl,
			Admin:        gf.Creator.Admin,
			Owner:        gf.Creator.Owner,
		}
	}

	return f
}

// forwardReplyFromGenerated converts a generated ForwardReply to our clean type.
func forwardReplyFromGenerated(gr generated.ForwardReply) ForwardReply {
	r := ForwardReply{
		Status:    gr.Status,
		CreatedAt: gr.CreatedAt,
		UpdatedAt: gr.UpdatedAt,
		Content:   gr.Content,
		Type:      gr.Type,
		URL:       gr.Url,
		AppURL:    gr.AppUrl,
	}

	if gr.Id != nil {
		r.ID = *gr.Id
	}

	if gr.Parent.Id != nil || gr.Parent.Title != "" {
		r.Parent = &Parent{
			ID:     derefInt64(gr.Parent.Id),
			Title:  gr.Parent.Title,
			Type:   gr.Parent.Type,
			URL:    gr.Parent.Url,
			AppURL: gr.Parent.AppUrl,
		}
	}

	if gr.Bucket.Id != nil || gr.Bucket.Name != "" {
		r.Bucket = &Bucket{
			ID:   derefInt64(gr.Bucket.Id),
			Name: gr.Bucket.Name,
			Type: gr.Bucket.Type,
		}
	}

	if gr.Creator.Id != nil || gr.Creator.Name != "" {
		r.Creator = &Person{
			ID:           derefInt64(gr.Creator.Id),
			Name:         gr.Creator.Name,
			EmailAddress: gr.Creator.EmailAddress,
			AvatarURL:    gr.Creator.AvatarUrl,
			Admin:        gr.Creator.Admin,
			Owner:        gr.Creator.Owner,
		}
	}

	return r
}

File: typescript/src/services/forwards.ts
/**
 * Forwards service for the Basecamp API.
 *
 * Forwards are emails that have been forwarded to a project's inbox.
 * Team members can reply to forwarded emails from within Basecamp.
 *
 * @example
 * ```ts
 * const inbox = await client.forwards.getInbox(projectId, inboxId);
 * const forwards = await client.forwards.list(projectId, inboxId);
 * const reply = await client.forwards.createReply(projectId, forwardId, {
 *   content: "<p>Thanks for reaching out!</p>",
 * });
 * ```
 */

import { BaseService } from "./base.js";
import { Errors } from "../errors.js";

// =============================================================================
// Types
// =============================================================================

/**
 * A person reference (simplified).
 */
export interface PersonRef {
  id: number;
  name: string;
  email_address?: string;
  avatar_url?: string;
  admin?: boolean;
  owner?: boolean;
}

/**
 * A bucket (project) reference.
 */
export interface BucketRef {
  id: number;
  name: string;
  type: string;
}

/**
 * A parent reference.
 */
export interface ParentRef {
  id: number;
  title: string;
  type: string;
  url: string;
  app_url: string;
}

/**
 * A Basecamp inbox (forwards tool).
 */
export interface Inbox {
  id: number;
  status: string;
  created_at: string;
  updated_at: string;
  title: string;
  type: string;
  url: string;
  app_url: string;
  bucket?: BucketRef;
  creator?: PersonRef;
}

/**
 * A forwarded email in Basecamp.
 */
export interface Forward {
  id: number;
  status: string;
  created_at: string;
  updated_at: string;
  subject: string;
  content: string;
  from: string;
  type: string;
  url: string;
  app_url: string;
  parent?: ParentRef;
  bucket?: BucketRef;
  creator?: PersonRef;
}

/**
 * A reply to a forwarded email.
 */
export interface ForwardReply {
  id: number;
  status: string;
  created_at: string;
  updated_at: string;
  content: string;
  type: string;
  url: string;
  app_url: string;
  parent?: ParentRef;
  bucket?: BucketRef;
  creator?: PersonRef;
}

/**
 * Request to create a reply to a forwarded email.
 */
export interface CreateForwardReplyRequest {
  /** Reply body in HTML (required) */
  content: string;
}

// =============================================================================
// Service
// =============================================================================

/**
 * Service for managing email forwards in Basecamp.
 */
export class ForwardsService extends BaseService {
  /**
   * Gets an inbox by ID.
   *
   * @param projectId - The project (bucket) ID
   * @param inboxId - The inbox ID
   * @returns The inbox
   * @throws BasecampError with code "not_found" if inbox doesn't exist
   *
   * @example
   * ```ts
   * const inbox = await client.forwards.getInbox(projectId, inboxId);
   * console.log(inbox.title);
   * ```
   */
  async getInbox(projectId: number, inboxId: number): Promise<Inbox> {
    const response = await this.request(
      {
        service: "Forwards",
        operation: "GetInbox",
        resourceType: "inbox",
        isMutation: false,
        projectId,
        resourceId: inboxId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/inboxes/{inboxId}", {
          params: { path: { projectId, inboxId } },
        })
    );

    return response as unknown as Inbox;
  }

  /**
   * Lists all forwards in an inbox.
   *
   * @param projectId - The project (bucket) ID
   * @param inboxId - The inbox ID
   * @returns Array of forwards
   *
   * @example
   * ```ts
   * const forwards = await client.forwards.list(projectId, inboxId);
   * forwards.forEach(f => console.log(f.subject, f.from));
   * ```
   */
  async list(projectId: number, inboxId: number): Promise<Forward[]> {
    const response = await this.request(
      {
        service: "Forwards",
        operation: "List",
        resourceType: "forward",
        isMutation: false,
        projectId,
        resourceId: inboxId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/inboxes/{inboxId}/forwards.json", {
          params: { path: { projectId, inboxId } },
        })
    );

    return (response ?? []) as Forward[];
  }

  /**
   * Gets a forward by ID.
   *
   * @param projectId - The project (bucket) ID
   * @param forwardId - The forward ID
   * @returns The forward
   * @throws BasecampError with code "not_found" if forward doesn't exist
   *
   * @example
   * ```ts
   * const forward = await client.forwards.get(projectId, forwardId);
   * console.log(forward.subject, forward.content);
   * ```
   */
  async get(projectId: number, forwardId: number): Promise<Forward> {
    const response = await this.request(
      {
        service: "Forwards",
        operation: "Get",
        resourceType: "forward",
        isMutation: false,
        projectId,
        resourceId: forwardId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/inbox_forwards/{forwardId}", {
          params: { path: { projectId, forwardId } },
        })
    );

    return response as unknown as Forward;
  }

  /**
   * Lists all replies to a forward.
   *
   * @param projectId - The project (bucket) ID
   * @param forwardId - The forward ID
   * @returns Array of replies
   *
   * @example
   * ```ts
   * const replies = await client.forwards.listReplies(projectId, forwardId);
   * replies.forEach(r => console.log(r.content));
   * ```
   */
  async listReplies(projectId: number, forwardId: number): Promise<ForwardReply[]> {
    const response = await this.request(
      {
        service: "Forwards",
        operation: "ListReplies",
        resourceType: "forward_reply",
        isMutation: false,
        projectId,
        resourceId: forwardId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/inbox_forwards/{forwardId}/replies.json", {
          params: { path: { projectId, forwardId } },
        })
    );

    return (response ?? []) as ForwardReply[];
  }

  /**
   * Gets a forward reply by ID.
   *
   * @param projectId - The project (bucket) ID
   * @param forwardId - The forward ID
   * @param replyId - The reply ID
   * @returns The reply
   * @throws BasecampError with code "not_found" if reply doesn't exist
   *
   * @example
   * ```ts
   * const reply = await client.forwards.getReply(projectId, forwardId, replyId);
   * console.log(reply.content);
   * ```
   */
  async getReply(projectId: number, forwardId: number, replyId: number): Promise<ForwardReply> {
    const response = await this.request(
      {
        service: "Forwards",
        operation: "GetReply",
        resourceType: "forward_reply",
        isMutation: false,
        projectId,
        resourceId: replyId,
      },
      () =>
        this.client.GET("/buckets/{projectId}/inbox_forwards/{forwardId}/replies/{replyId}", {
          params: { path: { projectId, forwardId, replyId } },
        })
    );

    return response as unknown as ForwardReply;
  }

  /**
   * Creates a reply to a forwarded email.
   *
   * @param projectId - The project (bucket) ID
   * @param forwardId - The forward ID
   * @param req - Reply creation parameters
   * @returns The created reply
   * @throws BasecampError with code "validation" if content is missing
   *
   * @example
   * ```ts
   * const reply = await client.forwards.createReply(projectId, forwardId, {
   *   content: "<p>Thanks for your email!</p>",
   * });
   * ```
   */
  async createReply(
    projectId: number,
    forwardId: number,
    req: CreateForwardReplyRequest
  ): Promise<ForwardReply> {
    if (!req.content) {
      throw Errors.validation("Reply content is required");
    }

    const response = await this.request(
      {
        service: "Forwards",
        operation: "CreateReply",
        resourceType: "forward_reply",
        isMutation: true,
        projectId,
        resourceId: forwardId,
      },
      () =>
        this.client.POST("/buckets/{projectId}/inbox_forwards/{forwardId}/replies.json", {
          params: { path: { projectId, forwardId } },
          body: {
            content: req.content,
          },
        })
    );

    return response as unknown as ForwardReply;
  }
}
